### PR TITLE
feat(builder): make the builder pod timeout configurable

### LINKED ISSUE #135: make the builder pod timeout configurable
After #133 is done, it will be hard-coded. Make it configurable in an env variable

### PR DESCRIPTION
**Pull request by aledbf**
closes #135

### CODE REVIEW

**Comment by aledbf** [comment]
@arschles please tell me the name to use for this 2 variables

**Comment by arschles** [comment]
@aledbf to complete #135, I'd like the timeout and tick durations to be specified as environment variables, instead of hard-coding them as constants here.

Regarding your question, the environment variables could be named `OBJECT_STORAGE_WAIT_DURATION` and `OBJECT_STORAGE_TICK_DURATION`.

Finally, on startup, the builder should check to make sure the value for the tick duration is strictly less than the value for the wait duration (otherwise, the git receive hook would always fail...)

**Comment by aledbf** [comment]
>  instead of hard-coding them as constants here

I put that as default if there's no env variables defined. It's that ok?

**Comment by arschles** [comment]
Yes, that's ok if they're just defaults. In the code, though, I don't see the check for the environment variables.

Also, can you provide a separate set of tick and timeout variables for pod start and polling object storage. For example:

- `BUILDER_POD_TICK_DURATION` and `BUILDER_POD_WAIT_DURATION` for `waitForPod` call
- `OBJECT_STORAGE_TICK_DURATION` and `OBJECT_STORAGE_WAIT_DURATION` for the `storage.ObjectExists` calls

Also, if you put these values into the [`Config` struct](https://github.com/deis/builder/blob/master/pkg/gitreceive/config.go#L7-L23) then you can specify the defaults there and the [config package](https://github.com/kelseyhightower/envconfig) will take care of setting them if missing. Example code:

```go
BuilderPodTickDurationMSec int `envconfig:"BUILDER_POD_TICK_DURATION" default:"100"`
BuilderPodWaitDurationMSec int `envconfig:"BUILDER_POD_WAIT_DURATION" default:"300000"` // 5 minutes
ObjectStorageTickDurationMSec int `envconfing:"OBJECT_STORAGE_TICK_DURATION" default:"500"`
ObjectStorageWaitDurationMSec int `envconfig:"OBJECT_STORAGE_WAIT_DURATION" default:"300000"` // 5 minutes
```

**Comment by aledbf** [comment]
ping @arschles

File: pkg/gitreceive/config.go [hunk 1]
```diff
@@ -29,3 +34,27 @@ func (c Config) App() string {
...
+// the end of an operation that involves the object storage
+func (c Config) ObjectStorageTickDuration() time.Duration {
+	return time.Duration(time.Duration(c.ObjectStorageTickDurationMSec) * time.Millisecond)
+}
+
+// ObjectStorageWaitDuration returns the maximum time to wait for the end of an
+// operation that involves the object storage
+func (c Config) ObjectStorageWaitDuration() time.Duration {
+	return time.Duration(time.Duration(c.ObjectStorageWaitDurationMSec) * time.Millisecond)
+}
```
Comment (arschles) [inline pkg/gitreceive/config.go]: :+1: for these `func`s!

File: boot.go [hunk 2]
```diff
@@ -59,6 +59,15 @@ func main() {
...
 					os.Exit(1)
 				}
+
+				// check for invalid ticks duration
+				if cnf.BuilderPodTickDurationMSec >= cnf.BuilderPodWaitDurationMSec {
+					cnf.BuilderPodTickDurationMSec = 100
+				}
+				if cnf.ObjectStorageTickDurationMSec >= cnf.ObjectStorageWaitDurationMSec {
+					cnf.ObjectStorageTickDurationMSec = 500
+				}
```
Comment (arschles) [inline boot.go]: Nitpick: can you put the `100` and the `500` into constants?

**Comment by arschles** [comment]
@aledbf other than my nitpick, LGTM

**Comment by smothiki** [comment]
Is this still in Progress ??

**Comment by arschles** [comment]
@smothiki good question - @aledbf I've labeled appropriately, since I'm assuming you're ready to go on this.

**Comment by aledbf** [comment]
@arschles yes, ready for review

**Comment by smothiki** [comment]
I was not reviewing it as it said IN progress.

**Comment by arschles** [comment]
@smothiki can you review now please?

**Comment by smothiki** [comment]
I was just asking as I gave an LGTM and the label says IN Progress. So just wondering If this is still in progress

**Comment by arschles** [comment]
@smothiki ok - any comments on the code?

File: boot.go [hunk 3]
```diff
@@ -59,6 +61,15 @@ func main() {
...
 					os.Exit(1)
 				}
+
+				// check for invalid ticks duration
+				if cnf.BuilderPodTickDurationMSec >= cnf.BuilderPodWaitDurationMSec {
+					cnf.BuilderPodTickDurationMSec = builderPodTick
+				}
+				if cnf.ObjectStorageTickDurationMSec >= cnf.ObjectStorageWaitDurationMSec {
+					cnf.ObjectStorageTickDurationMSec = objectStorageTick
+				}
```
Comment (arschles) [inline boot.go]: @aledbf can you pull this logic out and write a test for it?

**Comment by arschles** [comment]
@aledbf I took my LGTM out because I left a request for tests

**Comment by smothiki** [comment]
I don't have comments previously. But now you pointed at writing unit tests. Should I remove my LGTM or should the unit tests be a separate PR/issue

**Comment by arschles** [comment]
I'm ok with you leaving your LGTM on here. I'll review again when ready. Regarding unit tests, I'd prefer them to go into this PR (and, generally speaking, for all future builder PRs - even if it changes code that wasn't previously unit tested)

**Comment by smothiki** [comment]
+1

File: boot.go (hunk 3 above)
Comment (aledbf) [inline boot.go]: @arschles sure

**Comment by aledbf** [comment]
@arschles done

**Comment by aledbf** [comment]
ping @arschles

File: pkg/gitreceive/config_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			t.Fatalf("expected %v but %v was returned (%v)", eCase.podWait, cnf.BuilderPodWaitDurationMSec, tCase)
+		}
+		if cnf.ObjectStorageTickDurationMSec != eCase.storageTick {
+			t.Fatalf("expected %v but %v was returned (%v)", eCase.storageTick, cnf.ObjectStorageTickDurationMSec, tCase)
+		}
+		if cnf.ObjectStorageWaitDurationMSec != eCase.storageWait {
+			t.Fatalf("expected %v but %v was returned (%v)", eCase.storageWait, cnf.ObjectStorageWaitDurationMSec, tCase)
+		}
+	}
+}
```
Comment (arschles) [inline pkg/gitreceive/config_test.go]: :+1:

**Comment by arschles** [comment]
@aledbf LGTM, thanks for adding tests. @smothiki can you review once more?

**Pull request merged**